### ISSUE TITLE
Data problems

### ISSUE THREAD

**Issue by KNwbq** [yuh-yang/MBHT-KDD22#6]
Hi, I've found that the data does not match what is stated in the paper. And here are the problems:
1. In MBHT paper, the experiment results are based on ['Taobao', 'Retailrocket', 'IJCAI'], but the datasets you release are actually ['Tmall', 'Retailrocket', 'IJCAI'], which confused me.
2. Table 1 shows that Retailrocket has 3 behaviors and IJCAI has 4 behaviors, but for the dataset you release, Retailrocket has 4 behaviors and IJCAI has 5 behaviors.
3.  Noting that the behavior is represented by id, I would like to know the actual type of behavior of the behavior id, which is important for me.

Looking forward to your reply, sorry to bother you over the weekend.

**Comment by yuh-yang** [comment]
Hi! Here are some explanations:
1. the released 'Tmall' is actually 'Taobao' in the paper. The dataset is originally from Taobao.com, and we marked it wrongly. Accordingly, the 'IJCAI' dataset is the one collected from Tmall.com.

2. `0` is for beh_type masking. As mentioned in the paper, we should mask all target behaviors, which is `buy`. And this is done in the preprocessing.

3. You may refer to the raw data :)
[IJCAI](https://tianchi.aliyun.com/dataset/42)
[Taobao](https://tianchi.aliyun.com/dataset/649)
[Retailrocket](https://www.kaggle.com/datasets/retailrocket/ecommerce-dataset)

**Comment by KNwbq** [comment]
I accept explanation 1, but I'm still a little confused about explanation 2 and explanation 3.

Taking IJCAI as an example, the raw data has 4 types of behaviors, even if you block _**all**_ the target behaviors, the number of behavior types should remain 4 logically. The one condition I can think of is that you block only part of the target behaviors, thus increasing the number of behavior types to 5. 
In addition, I noticed that in the `reconstruct_train_data` function, you use 0 to mask the item as pos_Item, which has brought me some confusion. First of all, the items to be masked can be purchased items or clicked items or other items, because you are just randomly masking items. Secondly, if the masking operation has been done in the preprocessing, why mask the item again here?

As for question 3, I actually want to know the corresponding type of ID, for example, "0 means click, 1 means add to shopping cart, 2 means purchase, and 3 means add to favorites". Sorry, my expression is not up to standard.

Hoping for your reply.

**Comment by yuh-yang** [comment]
Hi! @KNwbq Thanks for the discussion.

We thoroughly re-checked the experimental settings to make clear the dataset and data preprocessing issues. Briefly, the key point is that in the latest experiments that we report, **for purchase behaviors, masking all of them has been replaced by a higher mask probability for them, due to gradient issues.** It is noticeable that if you mask all a certain single type of behaviors, the behavioral embedding to it cannot be properly learned. We'll correct the typo in the paper on Arxiv. To help understand, we give a data preprocessing flow as:

1. duplicate the sequences to enrich training data (say, 3 times)
2. mask purchase behaviors in a high rate (say, 0.8)
3. pack the dataset
4. while training, further add random masks in the sequence regardless of the behavior type to enrich the training instances across different behaviors

Note that we did not care about the mapping of all behavior types, resorting the relations that you need from the processed data can be tough. Therefore, you can just follow the flow above and reprocess your own dataset. :)

**Comment by KNwbq** [comment]
I fully understand. Thank you for your kind reply.
